### PR TITLE
adding atelier haven 65 and 80 keyboard pcb firmware,

### PR DESCRIPTION
**Pull request by CMMS-Freather**
adding AH haven 65 solder and hotswap firmware code
adding AH haven 80 solder and hotswap firmware code

## Description
moved eeconfig related code to keymap level, per @zvecr suggested

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by waffle87 — CHANGES_REQUESTED** [review]
Why do `haven65_hotswap` and `haven65_solder` need their own keyboard directories? As far as firmware goes, they're identical:
```diff
diff haven65_hotswap/config.h haven65_solder/config.h
6d5
<
diff haven65_hotswap/info.json haven65_solder/info.json
2c2
<     "keyboard_name": "HAVEN65_HOTSWAP",
---
>     "keyboard_name": "HAVEN65_Solder",
8c8
<         "pid": "0x4665",
---
>         "pid": "0x4666",
Common subdirectories: haven65_hotswap/keymaps and haven65_solder/keymaps
diff haven65_hotswap/readme.md haven65_solder/readme.md
1,3c1
< #Haven65 Hotswap
<
< ![haven 65 hotswap](https://i.imgur.com/M1l1DaO.png)
---
> #Haven65 Solder
4a3
> ![haven 65 solder](https://i.imgur.com/PRUfH5l.png)
11c10
<     make ah/haven65_hotswap:default
---
>     make ah/haven65_solder:default
15c14
<     make ah/haven65_hotswap:default:flash
---
>     make ah/haven65_solder:default:flash
```
Same goes for the `haven80_hotswap` and `haven80_solder`:
```diff
diff haven80_hotswap/config.h haven80_solder/config.h
1c1
< // Copyright 2023 Freather
---
> // Copyright 2022 Freather
4a5,9
>
>
>
> // generated by KBFirmware JSON to QMK Parser
> // https://noroadsleft.github.io/kbf_qmk_converter/
Common subdirectories: haven80_hotswap/keymaps and haven80_solder/keymaps
diff haven80_hotswap/readme.md haven80_solder/readme.md
1c1
< #Haven80 Hotswap
---
> #Haven80 Solder
3c3
< ![haven 80 hotswap](https://i.imgur.com/lXKf8Cih.png)
---
> ![haven 80 hotswap](https://i.imgur.com/TBXFqE5.png)
11c11
<     make ah/haven80_hotswap:default
---
>     make ah/haven80_solder:default
15c15
<     make ah/haven80_hotswap:default:flash
---
>     make ah/haven80_solder:default:flash
17c17
< ##Bootloader Atmel-DFU
---
> ##Bootloader Atmel-DFU
```

File: keyboards/ah/haven65/info.json [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        "alternating": true,
+        "breathing": true,
+        "christmas": true,
+        "knight": true,
+        "rainbow_mood": true,
+        "rainbow_swirl": true,
+        "rgb_test": true,
+        "snake": true,
+        "static_gradient": true,
+        "twinkle": true
```
Comment (drashna) [inline keyboards/ah/haven65/info.json:48]: Nit, readability:
```suggestion
            "alternating": true,
            "breathing": true,
            "christmas": true,
            "knight": true,
            "rainbow_mood": true,
            "rainbow_swirl": true,
            "rgb_test": true,
            "snake": true,
            "static_gradient": true,
            "twinkle": true
```

File: keyboards/ah/haven65/keymaps/via/keymap.c [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+// On Keyboard startup
+void keyboard_post_init_user(void) {
+    // Read custom menu variables from memory
+    eeconfig_read_user_datablock(&freather);
+}
+
+// Handle the data received by the keyboard from the VIA menus
+void freather_config_set_value(uint8_t *data) {
```
Comment (drashna) [inline keyboards/ah/haven65/keymaps/via/keymap.c:97]: All of this via stuff is repetitive, and could/should be moved to a common file.

Comment (CMMS-Freather) [reply keyboards/ah/haven65/keymaps/via/keymap.c:97]: > All of this via stuff is repetitive, and could/should be moved to a common file.

so like a .c file in the main keyboard folder?

**Comment by CMMS-Freather** [comment]
> Why do `haven65_hotswap` and `haven65_solder` need their own keyboard directories? As far as firmware goes, they're identical:
> 
> ```diff
> diff haven65_hotswap/config.h haven65_solder/config.h
> 6d5
> <
> diff haven65_hotswap/info.json haven65_solder/info.json
> 2c2
> <     "keyboard_name": "HAVEN65_HOTSWAP",
> ---
> >     "keyboard_name": "HAVEN65_Solder",
> 8c8
> <         "pid": "0x4665",
> ---
> >         "pid": "0x4666",
> Common subdirectories: haven65_hotswap/keymaps and haven65_solder/keymaps
> diff haven65_hotswap/readme.md haven65_solder/readme.md
> 1,3c1
> < #Haven65 Hotswap
> <
> < haven 65 hotswap
> ---
> > #Haven65 Solder
> 4a3
> > haven 65 solder
> 11c10
> <     make ah/haven65_hotswap:default
> ---
> >     make ah/haven65_solder:default
> 15c14
> <     make ah/haven65_hotswap:default:flash
> ---
> >     make ah/haven65_solder:default:flash
> ```
> 
> Same goes for the `haven80_hotswap` and `haven80_solder`:
> 
> ```diff
> diff haven80_hotswap/config.h haven80_solder/config.h
> 1c1
> < // Copyright 2023 Freather
> ---
> > // Copyright 2022 Freather
> 4a5,9
> >
> >
> >
> > // generated by KBFirmware JSON to QMK Parser
> > // https://noroadsleft.github.io/kbf_qmk_converter/
> Common subdirectories: haven80_hotswap/keymaps and haven80_solder/keymaps
> diff haven80_hotswap/readme.md haven80_solder/readme.md
> 1c1
> < #Haven80 Hotswap
> ---
> > #Haven80 Solder
> 3c3
> < haven 80 hotswap
> ---
> > haven 80 hotswap
> 11c11
> <     make ah/haven80_hotswap:default
> ---
> >     make ah/haven80_solder:default
> 15c15
> <     make ah/haven80_hotswap:default:flash
> ---
> >     make ah/haven80_solder:default:flash
> 17c17
> < ##Bootloader Atmel-DFU
> ---
> > ##Bootloader Atmel-DFU
> ```

this has been resolved by comibining the 2 file into 1. multilayout can be enabled in via

File: keyboards/ah/haven65/config.h [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// Copyright 2023 Freather
+// SPDX-License-Identifier: GPL-2.0-or-later
+
```
Comment (waffle87) [inline keyboards/ah/haven65/config.h:3]: This file should be removed.

File: keyboards/ah/haven80_hotswap/config.h [hunk 4]
```diff
@@ -0,0 +1,4 @@
+// Copyright 2023 Freather
+// SPDX-License-Identifier: GPL-2.0-or-later
+
```
Comment (waffle87) [inline keyboards/ah/haven80_hotswap/config.h:3]: This file should be removed.

File: keyboards/teahouse/ayleen/config.h [hunk 5]
```diff
@@ -4,18 +4,5 @@
 #pragma once
```
Comment (waffle87) [inline keyboards/teahouse/ayleen/config.h:6]: This file should be removed.

**Review by waffle87 — CHANGES_REQUESTED** [review]
Since it looks like `haven80_solder/hotswap` just differ in pins used, it would be preferable to have one parent directory with subfolders as such
```shell
.
├── hotswap
│   └── info.json # contains info specific to hotswap pcb
├── info.json # contains info that applies to both pcbs
├── keymaps
└── solder
    └── info.json # contains info specific to solder pcb
```

File: keyboards/ah/haven80_hotswap/config.h (hunk 4 above)
Comment (CMMS-Freather) [reply keyboards/ah/haven80_hotswap/config.h]: updated per requested

**Comment by CMMS-Freather** [comment]
ready for new review

File: haven80_hotswap.json [hunk 6]
```diff
@@ -0,0 +1,231 @@
+{
```
Comment (zvecr) [inline haven80_hotswap.json:1]: > Do not include VIA json files in the PR. These do not belong in the QMK repository as they are not used by QMK firmware – they belong in the [VIA Keyboard Repo](https://github.com/the-via/keyboards)

File: haven80_solder.json [hunk 7]
```diff
@@ -0,0 +1,269 @@
+{
```
Comment (zvecr) [inline haven80_solder.json:1]: Nope on this one too

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/ah/haven65/info.json [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        "bootmagic": true,
+        "mousekey": true,
+        "extrakey": true,
+        "console": false,
+        "command": true,
+        "nkro": true,
+        "backlight": false,
+        "rgblight": true,
+        "audio": false,
+        "key_lock": false
```
Comment (zvecr) [inline keyboards/ah/haven65/info.json:28]:
```suggestion
        "rgblight": true
```

File: keyboards/ah/haven65/info.json [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    },
+    "rgblight": {
+        "pin": "C7",
+        "led_count": 1,
+        "hue_steps": 8,
+        "saturation_steps": 8,
+        "brightness_steps": 8,
+        "max_brightness": 255,
```
Comment (zvecr) [inline keyboards/ah/haven65/info.json:36]:
```suggestion
```

File: keyboards/ah/haven65/info.json [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+              { "label": "space",           "matrix":[4,6],"x": 3.75, "y": 4, "w": 6.25 },
+              { "label": "ralt",            "matrix":[4,9],"x": 10, "y": 4, "w": 1.25 },
+              { "label": "MO(1)",           "matrix":[4,10],"x": 11.25, "y": 4, "w": 1.25 },
+              { "label": "left",            "matrix":[4,11],"x": 13, "y": 4 },
+              { "label": "down",            "matrix":[4,12],"x": 14, "y": 4 },
+              { "label": "right",           "matrix":[4,13],"x": 15, "y": 4 }
+            ]
+        }
+    },
+    "meta": "https://noroadsleft.github.io/kbf_qmk_converter/"
```
Comment (zvecr) [inline keyboards/ah/haven65/info.json:128]:
```suggestion
```

File: keyboards/ah/haven80/info.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        "bootmagic": true,
+        "mousekey": true,
+        "extrakey": true,
+        "console": false,
+        "command": false,
+        "nkro": true,
+        "backlight": false,
+        "rgblight": true,
+        "audio": false,
+        "key_lock": false
```
Comment (zvecr) [inline keyboards/ah/haven80/info.json:18]:
```suggestion
        "rgblight": true
```

File: keyboards/ah/haven80/solder/info.json [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        "bootmagic": true,
+        "mousekey": true,
+        "extrakey": true,
+        "console": false,
+        "command": false,
+        "nkro": true,
+        "backlight": false,
+        "rgblight": true,
+        "audio": false,
+        "key_lock": false
```
Comment (zvecr) [inline keyboards/ah/haven80/solder/info.json:23]:
```suggestion
        "rgblight": true
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Also via caps indicator stuff might now be rejected due to duplication/fragmentation within the repo.

**Comment by CMMS-Freather** [comment]
> Also via caps indicator stuff might now be rejected due to duplication/fragmentation within the repo.

is there anyway to make it fit? its just an custom ui function written like cipulot's ec pro2 custom ui function to adjust the threshold

**Comment by CMMS-Freather** [comment]
@waffle87 would you mind taking another look?

**Comment by CMMS-Freather** [comment]
@drashna @zvecr @waffle87 bump

**Comment by CMMS-Freather** [comment]
@zvecr @waffle87 @drashna bump

File: keyboards/ah/haven65/info.json [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    "features": {
+        "bootmagic": true,
+        "mousekey": true,
+        "extrakey": true,
+        "command": true,
+        "nkro": true,
+        "rgblight": true
+    },
+    "rgblight": {
+        "pin": "C7",
```
Comment (drashna) [inline keyboards/ah/haven65/info.json:27]:
```suggestion
    "ws2812": {
        "pin": "C7",
    },
    "rgblight": {
        "pin": "C7",
```

File: keyboards/ah/haven65/keymaps/via/config.h [hunk 14]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2023 Freather
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+
+#define EECONFIG_USER_DATA_SIZE 4
```
Comment (drashna) [inline keyboards/ah/haven65/keymaps/via/config.h:7]: The default eeconfig size is 4 (and a bit simpler).  So if you don't need more than this ...

File: keyboards/ah/haven65/keymaps/via/keymap.c [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+int mode = 5;
+
+//to change color if the caps lock is on
+void freather_set_color(bool capslock_is_on);
+bool process_record_user(uint16_t keycode, keyrecord_t *record);
+
+
+
+// Declaring a _freather_config_t struct that will store data
+typedef struct _indicator_config_t {
```
Comment (drashna) [inline keyboards/ah/haven65/keymaps/via/keymap.c:50]: this won't get merged with this indicator code.

File: keyboards/ah/haven65/keymaps/via/config.h (hunk 14 above)
Comment (CMMS-Freather) [reply keyboards/ah/haven65/keymaps/via/config.h]: this file has been removed

File: keyboards/ah/haven65/keymaps/via/keymap.c (hunk 15 above)
Comment (CMMS-Freather) [reply keyboards/ah/haven65/keymaps/via/keymap.c]: all parts related to via specific code block has been removed.
using led_update to implement the function of caps lock update for now

**Comment by CMMS-Freather** [comment]
the code has been updated to fit the qmk standard (removing the via relatated code block) and tested working on boards.

**Comment by phillmund** [comment]
Is this such a big issue that it takes so long to merge the branch? What can I do to speed up the process?

**Comment by CMMS-Freather** [comment]
@zvecr @drashna @waffle87 

the checks have passed and formatted in the required format. please review and approve if possible

File: keyboards/ah/haven65/keymaps/default/keymap.c [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    bool res = led_update_user(led_state);
+    if (res) {
+    if (led_state.caps_lock) {
+        rgblight_sethsv_at(0, 0, 255, 0);
+    } else {
+        rgblight_sethsv_at(0, 0, 0, 0);
+    }
+  }
+    return res;
+}
```
Comment (waffle87) [inline keyboards/ah/haven65/keymaps/default/keymap.c:34]: This should be placed in a `haven65.c` file.
```suggestion
```

File: keyboards/ah/haven65/keymaps/via/keymap.c [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    bool res = led_update_user(led_state);
+    if (res) {
+    if (led_state.caps_lock) {
+        rgblight_sethsv_at(0, 0, 255, 0);
+    } else {
+        rgblight_sethsv_at(0, 0, 0, 0);
+    }
+  }
+    return res;
+}
```
Comment (waffle87) [inline keyboards/ah/haven65/keymaps/via/keymap.c:48]:
```suggestion
  )
};
```

File: keyboards/ah/haven80/keymaps/default/keymap.c [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+    bool res = led_update_user(led_state);
+    if (res) {
+    if (led_state.caps_lock) {
+        rgblight_sethsv_at(0, 0, 255, 0); 
+    } else {
+        rgblight_sethsv_at(0, 0, 0, 0);
+    }
+  }
+    return res;
+}
```
Comment (waffle87) [inline keyboards/ah/haven80/keymaps/default/keymap.c:36]: This should be placed in a `haven80.c` file.
```suggestion
```

File: keyboards/ah/haven80/keymaps/via/keymap.c [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    bool res = led_update_user(led_state);
+    if (res) {
+    if (led_state.caps_lock) {
+        rgblight_sethsv_at(0, 0, 255, 0);
+    } else {
+        rgblight_sethsv_at(0, 0, 0, 0);
+    }
+  }
+    return res;
+}
```
Comment (waffle87) [inline keyboards/ah/haven80/keymaps/via/keymap.c:52]:
```suggestion
  )
};
```

File: keyboards/ah/haven80/solder/readme.md [hunk 20]
```diff
@@ -0,0 +1,21 @@
+#Haven80 Solder
```
Comment (waffle87) [inline keyboards/ah/haven80/solder/readme.md:1]:
```suggestion
# Haven80 Solder
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/teahouse/ayleen/keymaps/default/keymap.c [hunk 21]
```diff
@@ -12,5 +12,25 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
...
         KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT,  KC_UP,
         KC_LCTL, KC_LGUI, KC_LALT, KC_SPC,  KC_RALT, KC_RGUI, KC_RCTL, KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+    [1] = LAYOUT(
+        KC_SPC,  _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______
```
[the remark this answers is not in the document]
Comment (CMMS-Freather) [reply keyboards/teahouse/ayleen/keymaps/default/keymap.c]: this part was in my old understading that in the default keyamp, there needs at least 2 layers, and in via keymap, there needs at least 4 layers to be defined. if this is not needed nowdays, i will remove

**Comment by CMMS-Freather** [comment]
@waffle87 I have updated the code as requested, however, in the pr link keyboards, it says error code 1, do you have any suggestions on what caused it? since i wasn't able to see the reason why the error happened

**Comment by CMMS-Freather** [comment]
@waffle87 bump

File: keyboards/ah/haven65/readme.md [hunk 22]
```diff
@@ -0,0 +1,21 @@
+#Haven65
+
+![haven 65](https://i.imgur.com/M1l1DaO.png)
```
Comment (drashna) [inline keyboards/ah/haven65/readme.md:3]:
```suggestion
![haven 65](https://i.imgur.com/M1l1DaOh.png)
```

File: keyboards/ah/haven80/keymaps/via/keymap.c [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______
+  ),
+  [3] = LAYOUT(
+      KC_SPC,  _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______
+  )
```
Comment (drashna) [inline keyboards/ah/haven80/keymaps/via/keymap.c:39]:
```suggestion
```

the dynamic keymap feature will auto-fill any missing layer.

File: keyboards/ah/haven80/solder/readme.md [hunk 24]
```diff
@@ -0,0 +1,21 @@
+# Haven80 Solder
+
+![haven 80 hotswap](https://i.imgur.com/TBXFqE5.png)
```
Comment (drashna) [inline keyboards/ah/haven80/solder/readme.md:3]:
```suggestion
![haven 80 hotswap](https://i.imgur.com/TBXFqE5h.png)
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by CMMS-Freather** [comment]
@drashna bump

**Review by drashna — APPROVED** [review]

**Comment by waffle87** [comment]
There are still unresolved suggestions...
eg. [here](https://github.com/qmk/qmk_firmware/pull/20142#discussion_r1234212433)

**Review by waffle87 — APPROVED** [review]

**Pull request merged**